deepspeed/runtime/zero/stage3.py

```diff
@@ -1169,15 +1168,13 @@ def create_reduce_and_remove_grad_hooks(self):
 
                     #print(f"After all gather {param.device}, {param.shape}")
                     def wrapper(param):
-                        param_tmp = param.expand_as(param)
-                        grad_acc = param_tmp.grad_fn.next_functions[0][0]
 
                         @instrument_w_nvtx
                         def reduce_partition_and_remove_grads(*notneeded):
                             self.reduce_ready_partitions_and_remove_grads(param)
 
-                        self._grad_acc_hooks.append(grad_acc.register_hook(reduce_partition_and_remove_grads))
-                        self.grad_accs.append(grad_acc)
+                        self._grad_acc_hooks.append(
+                            param.register_post_accumulate_grad_hook(reduce_partition_and_remove_grads))
```

Comment:
Which pytorch version introduced this API? How should we handle older versions?

Reply:
Hi @tjruwase

- This API was introduced starting pytorch v2.1.
- We can add check what version and use the older API when needed.
- However, compilation should be disabled for the older versions to avoid the accuracy issues (if activation checkpointing is enabled).

Please advice what should be the approach. Thanks.

Reply:
In fact, in PT 2.0 the compile feature was still quite unstable. So, maybe we discourage compilation on version earlier than 2.1 by displaying a warning in the engine.compile().
@tjruwase what are your thoughts on this?

Reply:
@deepcharm, apologies for the delayed response. I am aligned with your suggestion, specifically, we should enable for >= 2.1 and disable for older versions. In other words, no need to warn for 2.0 given the unstable nature. Thanks!

Reply:
@tjruwase thanks for your comments. I've updated the patch which now:
1) Uses proper grad accumulation hook API depending on the PyTorch version
2) Disables compile for PyTorch < 2.1

Let me know if anything is missing. Thanks

Reply:
@deepcharm, looks great. Thanks!